decorate/index.ts

```diff
@@ -0,0 +1,212 @@
+  setHeader?: (
+    name: string,
+    value: number | string | ReadonlyArray<string>,
+  ) => unknown;
+}
+
+function isResponseLike(response: ArcjetResponse): response is ResponseLike {
+  if (typeof response.headers === "undefined") {
+    return false;
+  }
+
+  if (
+    "has" in response.headers &&
+    typeof response.headers.has === "function" &&
+    "set" in response.headers &&
+    typeof response.headers.set === "function"
+  ) {
+    return true;
+  }
+
+  return false;
+}
+
+function isOutgoingMessageLike(
+  response: ArcjetResponse,
+): response is OutgoingMessageLike {
+  if (typeof response.headersSent !== "boolean") {
+    return false;
+  }
+
+  if (typeof response.hasHeader !== "function") {
+    return false;
+  }
+
+  if (typeof response.setHeader !== "function") {
+    return false;
+  }
+
+  return true;
+}
+
+function sortByLowestMax(
+  [maxA, _windowA]: [number, number],
+  [maxB, _windowB]: [number, number],
+) {
+  return maxA - maxB;
+}
+
+function toPolicyString([max, window]: [number, number]) {
+  return `${max};w=${window}`;
+}
+
+function extractReason(result: ArcjetRuleResult): ArcjetReason {
+  return result.reason;
+}
+
+function isRateLimitReason(
+  reason: ArcjetReason,
+): reason is ArcjetRateLimitReason {
+  return reason.isRateLimit();
+}
+
+function nearestLimit(
+  current: ArcjetRateLimitReason,
+  next: ArcjetRateLimitReason,
+) {
+  if (current.remaining < next.remaining) {
+    return current;
+  }
+
+  if (current.remaining > next.remaining) {
+    return next;
+  }
+
+  // Reaching here means `remaining` is equal so prioritize closest reset
+  if (current.reset < next.reset) {
+    return current;
+  }
+
+  if (current.reset > next.reset) {
+    return next;
+  }
+
+  // Reaching here means that `remaining` and `reset` are equal, so prioritize
+  // the smallest `max`
+  if (current.max < next.max) {
+    return current;
+  }
+
+  if (current.max > next.max) {
+    return next;
+  }
+
+  // If all else is equal, just return the current reason
+  return current;
+}
+
+/**
+ * Decorates a response with `RateLimit` and `RateLimit-Policy` headers based
+ * on an {@link ArcjetDecision} and conforming to the [Rate Limit fields for
+ * HTTP](https://ietf-wg-httpapi.github.io/ratelimit-headers/draft-ietf-httpapi-ratelimit-headers.html)
+ * draft specification.
+ *
+ * @param response The response to decorate—must be similar to a DOM Response or node's OutgoingMessage.
+ * @param decision The {@link ArcjetDecision} that was made by calling `protect()` on the SDK.
+ */
+export function setRateLimitHeaders(
+  response: ArcjetResponse,
+  decision: ArcjetDecision,
+) {
+  const rateLimitReasons = decision.results
+    .map(extractReason)
+    .filter(isRateLimitReason);
+
+  const policies = new Map<number, number>();
+  for (const reason of rateLimitReasons) {
+    if (policies.has(reason.max)) {
+      // TODO: Warn? Or does it make sense to bail since there is conflicting policies?
```

Comment:
I suggest log a warning, but then set them anyway. The developer should know what they want when calling this. It's easier for us to become more strict e.g. throwing, later on that it will be to go the other way.

Reply:
Updated. There are a couple cases where we should return, but those are now `logger.error`